covenant-signer/itest/containers/containers.go

```diff
@@ -0,0 +1,187 @@
+	if err != nil {
+		return nil, err
+	}
+	return docker, nil
+}
+
+func (m *Manager) ExecBitcoindCliCmd(t *testing.T, command []string) (bytes.Buffer, bytes.Buffer, error) {
+	// this is currently hardcoded, as it will be the same for all tests
+	cmd := []string{"bitcoin-cli", "-chain=regtest", "-rpcuser=user", "-rpcpassword=pass"}
+	cmd = append(cmd, command...)
+	return m.ExecCmd(t, bitcoindContainerName, cmd)
+}
+
+// ExecCmd executes command by running it on the given container.
+// It word for word `error` in output to discern between error and regular output.
+// It retures stdout and stderr as bytes.Buffer and an error if the command fails.
+func (m *Manager) ExecCmd(t *testing.T, containerName string, command []string) (bytes.Buffer, bytes.Buffer, error) {
+	if _, ok := m.resources[containerName]; !ok {
+		return bytes.Buffer{}, bytes.Buffer{}, fmt.Errorf("no resource %s found", containerName)
+	}
+	containerId := m.resources[containerName].Container.ID
+
+	var (
+		outBuf bytes.Buffer
+		errBuf bytes.Buffer
+	)
+
+	timeout := 20 * time.Second
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	t.Logf("\n\nRunning: \"%s\"", command)
+
+	// We use the `require.Eventually` function because it is only allowed to do one transaction per block without
+	// sequence numbers. For simplicity, we avoid keeping track of the sequence number and just use the `require.Eventually`.
+	require.Eventually(
+		t,
+		func() bool {
+			exec, err := m.pool.Client.CreateExec(docker.CreateExecOptions{
+				Context:      ctx,
+				AttachStdout: true,
+				AttachStderr: true,
+				Container:    containerId,
+				User:         "root",
+				Cmd:          command,
+			})
+
+			if err != nil {
+				t.Logf("failed to create exec: %v", err)
+				return false
+			}
+
+			err = m.pool.Client.StartExec(exec.ID, docker.StartExecOptions{
+				Context:      ctx,
+				Detach:       false,
+				OutputStream: &outBuf,
+				ErrorStream:  &errBuf,
+			})
+			if err != nil {
+				t.Logf("failed to start exec: %v", err)
+				return false
+			}
+
+			errBufString := errBuf.String()
+			// Note that this does not match all errors.
+			// This only works if CLI outputs "Error" or "error"
+			// to stderr.
+			if errRegex.MatchString(errBufString) {
+				t.Log("\nstderr:")
+				t.Log(errBufString)
+
+				t.Log("\nstdout:")
+				t.Log(outBuf.String())
+				return false
+			}
+
+			return true
+		},
+		timeout,
+		500*time.Millisecond,
+		"command failed",
+	)
+
+	return outBuf, errBuf, nil
+}
+
+func (m *Manager) RunBitcoindResource(
+	bitcoindCfgPath string,
+) (*dockertest.Resource, error) {
+	bitcoindResource, err := m.pool.RunWithOptions(
+		&dockertest.RunOptions{
+			Name:       bitcoindContainerName,
+			Repository: m.cfg.BitcoindRepository,
+			Tag:        m.cfg.BitcoindVersion,
+			User:       "root:root",
+			Mounts: []string{
+				fmt.Sprintf("%s/:/data/.bitcoin", bitcoindCfgPath),
+			},
+			ExposedPorts: []string{
+				"8332",
```

Comment:
FYI: these should have a suffix `/tcp`, but because the actually dockerfile image exposes them this works, otherwise it is ignored